### PR TITLE
Feat/142 collection stats

### LINKED ISSUE #142: Resolver for stats of collection
Requirement
- extends `collectionEntitiesConnection` resolver or provide a new resolver
- input:
```javascript
{
    offset
    limit
    where
    orderBy
}
```

- output:
```javascript
{
   volume:  xxx, // Total trading turnover
   highestSale: xxx, // Maximum selling price
   totalCount: xxx     // count of nfts in this colleciton
}
```

image

### PR DESCRIPTION
**Pull request by alko89**
**Thank you for your contribution** to the [KodaDot Indexer](https://kodadot.xyz).
👇 \_ Let's make a quick check before the contribution.

### PR type

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring

### What's new?

- [x] PR closes #142
- [x] This PR is required for kodadot/nft-gallery#5032

### Before submitting Pull Request, please make sure:

- [ ] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [ ] I've didn't break any original functionality
- [ ] I've posted a screenshot of demonstrated change in this PR

### Optional

- [ ] I've run the indexer and it hasn't failed
- [x] I've changed schema and performed a migration
- [ ] I found edge cases

### Screenshot

- [ ] My contribution is a **resolver** so I have pic from playground

### CODE REVIEW

**Comment by alko89** [comment]
@vikiival lmk if this is ok, then I can also charge for https://github.com/kodadot/loligo/issues/10

**Comment by vikiival** [comment]
Hello please check the PR that was opened by @matehoo
- https://github.com/kodadot/rubick/pull/236

File: src/mappings/index.ts [hunk 1]
```diff
@@ -248,6 +250,10 @@ export async function handleTokenBuy(context: Context): Promise<void> {
 
   const collection = ensure<CE>(await get<CE>(context.store, CE, event.collectionId));
   plsBe(real, collection);
+  collection.volume++;
```
Comment (vikiival) [inline src/mappings/index.ts:253]:
```suggestion
  collection.volume += event.price;
```

**Comment by alko89** [comment]
> Hello please check the PR that was opened by @Matehoo
> 
>     * [feat: Add stats for collections rubick#236](https://github.com/kodadot/rubick/pull/236)

Ohh didn't notice there should also be `distribution`, `floor` and `ownerCount`. Was wondering where this came from and noticed it mentioned in https://github.com/kodadot/loligo/issues/10 :sweat_smile: I'll update

**Comment by vikiival** [comment]
> > Hello please check the PR that was opened by @Matehoo
> 
> Ohh didn't notice there should also be `distribution`, `floor` and `ownerCount`. Was wondering where this came from and noticed it mentioned in [kodadot/loligo#10](https://github.com/kodadot/loligo/issues/10) 😅 I'll add them tomorrow

No worries, everything should be explained there. If something is not clear happy to explain

File: src/mappings/index.ts [hunk 2]
```diff
@@ -156,6 +160,14 @@ export async function handleTokenCreate(context: Context): Promise<void> {
...
   collection.supply += 1;
 
+  const { ownerCount, distribution } = await calculateCollectionOwnerCountAndDistribution(
+    context.store,
+    final.id,
+    final.currentOwner,
+  );
+  collection.ownerCount = ownerCount;
+  collection.distribution = distribution;
+
```
Comment (vikiival) [inline src/mappings/index.ts:170]: mint won't change much distribution imo

Comment (alko89) [reply src/mappings/index.ts:170]: can only the owner of the collection mint new NFTs?

Comment (vikiival) [reply src/mappings/index.ts:170]: yup

Comment (alko89) [reply src/mappings/index.ts:170]: ahh ok, sorry about that

Comment (vikiival) [reply src/mappings/index.ts]: Do not worry will try to make docs.  🤗

**Comment by Matehoo** [comment]
@vikiival @alko89 What is the status on this MR?

File: src/mappings/utils/entity.ts [hunk 3]
```diff
@@ -72,3 +72,37 @@ export function create<T extends EntityWithId>(
...
+              END)
+   FROM nft_entity) AS adjustment
+  `
+} 
+  FROM nft_entity
+  WHERE collection_id = '${collectionId}'
+  ${newOwner && `AND current_owner != '${originalOwner}'`}
+  `;
+  const queryResult: { owner_count: number; distribution: number; adjustment?: number }[] = await store.query(query);
+  const result = queryResult[0];
```
Comment (Matehoo) [inline src/mappings/utils/entity.ts:100]: This is something that I forgot to refactor
```suggestion
  const [result]: { owner_count: number; distribution: number; adjustment?: number }[] = await store.query(query);
```

File: src/server-extension/resolvers/index.ts [hunk 4]
```diff
@@ -4,8 +4,10 @@ import { OfferStatsResolver } from './offerStatsResolver';
 import { EventResolver } from './event';
 import { SeriesResolver } from './series';
 import { SpotlightResolver } from './spotlight';
+// import { CollectionResolver } from './collectionResolver';
```
Comment (Matehoo) [inline src/server-extension/resolvers/index.ts:7]: Should remove unused imports

File: src/server-extension/resolvers/index.ts [hunk 5]
```diff
@@ -4,8 +4,10 @@ import { OfferStatsResolver } from './offerStatsResolver';
...
 
 export {
+  // CollectionResolver,
```
Comment (Matehoo) [inline src/server-extension/resolvers/index.ts:10]: Same as the above

**Review by Matehoo — APPROVED** [review]
Some tiny cosmetic stuff, LGTM

File: src/server-extension/resolvers/index.ts (hunk 4 above)
Comment (alko89) [reply src/server-extension/resolvers/index.ts:7]: ohh good catch :+1: forgot to save file after removing, was doing some testing

File: src/model/generated/nftEntity.model.ts [hunk 6]
```diff
@@ -7,6 +7,9 @@ import {MetadataEntity} from "./metadataEntity.model"
 
 @Entity_()
 export class NFTEntity {
+    static createQueryBuilder(arg0: string) {
+      throw new Error('Method not implemented.')
+    }
```
Comment (vikiival) [inline src/model/generated/nftEntity.model.ts:12]: was this generated by squid @alko89 ?

**Review by vikiival — APPROVED** [review]

**Comment by Matehoo** [comment]
@vikiival @alko89 ETA on merge?

**Comment by yangwao** [comment]
@vikiival

**Review by roiLeo — CHANGES_REQUESTED** [review]
Doesn't work for me on Rococo

Here is log output:
```bash
✔  success   [BURN] 3647975502-8 by bXjMVEdCzhjYbPc5rJk7NvPSDRiiNuCT3GdqohFp3wdpzJL6Z
FATAL sqd:processor QueryFailedError: syntax error at or near "undefined"
  at PostgresQueryRunner.query (.../snek/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:211:19)
  at processTicksAndRejections (node:internal/process/task_queues:96:5)
  at async DataSource.query (.../snek/node_modules/typeorm/data-source/DataSource.js:341:20)
  at async calculateCollectionOwnerCountAndDistribution (.../snek/lib/mappings/utils/entity.js:63:22)
  at async handleTokenBurn (.../snek/lib/mappings/index.js:188:42)
  at async HandlerRunner.processBlock (.../snek/node_modules/@subsquid/substrate-processor/lib/processor/handlerProcessor.js:505:25)
  at async .../snek/node_modules/@subsquid/typeorm-store/lib/database.js:157:13
  at async EntityManager.transaction (.../snek/node_modules/typeorm/entity-manager/EntityManager.js:72:28)
  at async FullTypeormDatabase.runTransaction (.../snek/node_modules/@subsquid/typeorm-store/lib/database.js:155:9)
  at async FullTypeormDatabase.transact (.../snek/node_modules/@subsquid/typeorm-store/lib/database.js:64:24)
  query:
          SELECT COUNT(DISTINCT current_owner) AS distribution,
                COUNT(current_owner) AS owner_count
          undefined
          FROM nft_entity
          WHERE collection_id = '3647975502-8'
          undefined

  driverError: {"length":99,"name":"error","severity":"ERROR","code":"42601","position":"103","file":"scan.l","line":"1176","routine":"scanner_yyerror","stack":"error: syntax error at or near \"undefined\"\n    at Parser.parseErrorMessage (.../snek/node_modules/pg-protocol/dist/parser.js:287:98)\n    at Parser.handlePacket (.../snek/node_modules/pg-protocol/dist/parser.js:126:29)\n    at Parser.parse (.../snek/node_modules/pg-protocol/dist/parser.js:39:38)\n    at Socket.<anonymous> (.../snek/node_modules/pg-protocol/dist/index.js:11:42)\n    at Socket.emit (node:events:513:28)\n    at addChunk (node:internal/streams/readable:315:12)\n    at readableAddChunk (node:internal/streams/readable:289:9)\n    at Socket.Readable.push (node:internal/streams/readable:228:10)\n    at TCP.onStreamRead (node:internal/stream_base_commons:190:23)"}
  length: 99
  severity: ERROR
  code: 42601
  position: 103
  file: scan.l
  line: 1176
  routine: scanner_yyerror
  blockHeight: 750082
  blockHash: 0x429b514c452a93b57db272bcc16ebe353c1edbf23a0181a6abcac9aed4f89de2
make: *** [process] Error 1
```

**Comment by vikiival** [comment]
>  I've run the indexer and it hasn't failed

@alko89 please make sure that it runs

File: src/mappings/utils/entity.ts [hunk 7]
```diff
@@ -72,3 +72,36 @@ export function create<T extends EntityWithId>(
...
+                  WHEN current_owner = '${newOwner}' THEN 0
+                  ELSE 1
+              END)
+   FROM nft_entity) AS adjustment
+  `
+} 
+  FROM nft_entity
+  WHERE collection_id = '${collectionId}'
+  ${newOwner && `AND current_owner != '${originalOwner}'`}
+  `;
```
Comment (roiLeo) [inline src/mappings/utils/entity.ts:98]: This is difficult to read, I suggest this workaround (but not so pretty)
```suggestion
  let query;
  if (newOwner && originalOwner) {
    query = `SELECT 
      COUNT(DISTINCT current_owner) AS distribution,
      COUNT(current_owner) AS owner_count,
      (
        SELECT max(
          CASE
            WHEN current_owner = '${newOwner}' THEN 0
            ELSE 1
          END
        )
        FROM nft_entity
      ) AS adjustment
      FROM nft_entity
      WHERE collection_id = '${collectionId}'
      AND current_owner != '${originalOwner}'`;
  }

  query = `SELECT 
      COUNT(DISTINCT current_owner) AS distribution,
      COUNT(current_owner) AS owner_count
      FROM nft_entity
      WHERE collection_id = '${collectionId}'`;
```

Comment (vikiival) [reply src/mappings/utils/entity.ts:98]: pfff leeeet me fix that

Comment (roiLeo) [reply src/mappings/utils/entity.ts:98]: new event version?

Comment (vikiival) [reply src/mappings/utils/entity.ts:98]: Fixed in  #173

Comment (alko89) [reply src/mappings/utils/entity.ts]: @roiLeo did you maybe forgot to rebuild? I've just run `make process` after building and didn't encounter the error. I applied your suggestion.

**Review by roiLeo — APPROVED** [review]
`sqd:processor 1956220 / 1956220, rate: 5786 blocks/sec, mapping: 5 blocks/sec, 13 items/sec, ingest: 22 blocks/sec, eta: 0s`

**Comment by alko89** [comment]
@roiLeo fixed, used event id instead of collection id in the query

**Pull request merged**